### ISSUE TITLE
Policy-based auth

### ISSUE THREAD

**Issue by Nhowka** [giraffe-fsharp/Giraffe#224]
Looking at the asp.net docs, I found that feature could be useful. https://docs.microsoft.com/en-us/aspnet/core/security/authorization/policies

Is there a way to use it from Giraffe like the Roles?

**Comment by dustinmoris** [comment]
Hi, policy based auth is on my list. I am aiming to have this ready for the 1.2.0 release. 1.1.0 is aimed at going out latest by the end of this week just to give you some time frame.

**Comment by Nhowka** [comment]
After adding `Microsoft.AspNetCore.Authorization` to the project, I could add the authorization services using the code from [this article](https://vtquan.github.io/fsharp/setting-up-services-with-giraffe/). Then I created the method requiresPolicy as the following:

```F#
let requiresPolicy (policy : string) (authFailedHandler : HttpHandler) : HttpHandler =
    fun (next : HttpFunc) (ctx : HttpContext) ->                
        match ctx.GetService<IAuthorizationService>() with
        | null ->                     
            authFailedHandler finish ctx
        | authService -> task {            
            let! authResult = authService.AuthorizeAsync (ctx.User, policy)
            if authResult.Succeeded then
                return! next ctx
            else
                return! authFailedHandler finish ctx
            }
```

I tried to change the tests to use it, but I figured after a long time that it was not using the real deal and the errors were because the service was not injected. It worked in a different project, though.

**Comment by dustinmoris** [comment]
This has been released with [Giraffe 2.0.0](https://github.com/giraffe-fsharp/Giraffe/releases/tag/v2.0.0).
